<a href="https://colab.research.google.com/github/muzaffarizamuddin/Data-Science-Project/blob/main/P166246_Week13_DataAggregation_20260103.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Week 13: Data Aggregation and Group Operation**
Please refer to the textbook: **"Python for Data Analysis"** by Wes McKinney for details of this topic on **Chapter 10**

# **Start by importing NumPy and pandas**

In [ ]:
import numpy as np
import pandas as pd

# **13.1 Group Operations**
## The popular ***split-apply-combine***

In [ ]:
# Example DataFrame
rng = np.random.default_rng(seed=12345)
df = pd.DataFrame({"key1" : ["a", "a", None, "b", "b", "a", None],
                   "key2" : pd.Series([1, 2, 1, 2, 1, None, 1], dtype="Int64"),
                   "data1" : rng.standard_normal(7),
                   "data2" : rng.standard_normal(7)})
df

key1  key2     data1     data2
0     a     1 -1.423825  0.648893
1     a     2  1.263728  0.361058
2  None     1 -0.870662 -1.952863
3     b     2 -0.259173  2.347410
4     b     1 -0.075343  0.968497
5     a  <NA> -0.740885 -0.759387
6  None     1 -1.367793  0.902198

## Split -> groupby -> will create **intermediate grouped object**
## apply -> certain function
## combine -> get final output

In [ ]:
# Access data1 and call groupby with the key1 column
# Create an intermediate grouped object -> wanted to do sth else
grouped = df.groupby("key1")["data1"]

In [ ]:
# Compute mean
# If output has sth: "dtype: float64" -> Series
grouped.mean()

key1
a   -0.300327
b   -0.167258
Name: data1, dtype: float64

**By default**, pandas will:
1. **Sort group labels alphabetically**
2. **Exclude rows where the grouping key is missing (`NaN`)**
3. **Return the grouping key as the index**

## **Explicitly controlling GroupBy behavior**

To gain full control over the output, pandas provides three important parameters:

- **`sort=False`** → preserves the original order of appearance
    - By default, pandas **sorts group labels**, which may **reorder our results**

- **`dropna=False`** → includes missing values as a valid group
    - Missing values are frequently **informative**, especially in real-world datasets.

- **`as_index=False`** → keeps grouping keys as regular columns

In [ ]:
# Group by 'key1' and compute the mean of 'data1'
# while preserving row order, keeping missing keys, and returning a DataFrame
# Without intermediate grouped object
df.groupby(
    "key1",
    sort=False,
    dropna=False,
    as_index=False
)["data1"].mean()

key1     data1
0    a -0.300327
1  NaN -1.119227
2    b -0.167258

In [ ]:
# Passing multiple groupby arrays of keys to find mean
means = df["data1"].groupby([df["key1"], df["key2"]]).mean()
means

key1  key2
a     1      -1.423825
      2       1.263728
b     1      -0.075343
      2      -0.259173
Name: data1, dtype: float64

In [ ]:
# Unstack: from vertical to wide table
means.unstack()

key2         1         2
key1                    
a    -1.423825  1.263728
b    -0.075343 -0.259173

In [ ]:
# Pass column names as the group keys
# i want to groupby all the data using key1 as the label
df.groupby("key1").mean()

key2     data1     data2
key1                          
a      1.5 -0.300327  0.083521
b      1.5 -0.167258  1.657953

In [ ]:
# Groupby using multiple keys -> by providing a list
df.groupby(["key1", "key2"]).mean()

data1     data2
key1 key2                    
a    1    -1.423825  0.648893
     2     1.263728  0.361058
b    1    -0.075343  0.968497
     2    -0.259173  2.347410

In [ ]:
df

key1  key2     data1     data2
0     a     1 -1.423825  0.648893
1     a     2  1.263728  0.361058
2  None     1 -0.870662 -1.952863
3     b     2 -0.259173  2.347410
4     b     1 -0.075343  0.968497
5     a  <NA> -0.740885 -0.759387
6  None     1 -1.367793  0.902198

In [ ]:
# Returns a Series containing group sizes
df.groupby(["key1", "key2"]).size()

key1  key2
a     1       1
      2       1
b     1       1
      2       1
dtype: int64

In [ ]:
# Missing values in a group key are excluded from the result by default
# Can be disabled by passing dropna=False to groupby
# dropna=True is by default
df.groupby("key1").size()

key1
a    3
b    2
dtype: int64

In [ ]:
# Using dropna=False
df.groupby(["key1", "key2"], dropna=False).size()

key1  key2
a     1       1
      2       1
      <NA>    1
b     1       1
      2       1
NaN   1       2
dtype: int64

In [ ]:
# Can use count as well
df.groupby(["key1", "key2"])["data1"].count()

key1  key2
a     1       1
      2       1
b     1       1
      2       1
Name: data1, dtype: int64

In [ ]:
# Returns a Series containing group sizes
df.groupby(["key1", "key2"]).size()

key1  key2
a     1       1
      2       1
b     1       1
      2       1
dtype: int64

### **`count()`**: Returns the number of non-null values in each group **for each column**.

In [ ]:
# Difference between size() and count()
df = pd.DataFrame({"key1" : ["a", "a", None, "b", "b", "a", None],
                   "key2" : pd.Series([1, 2, 1, 2, 1, None, 1], dtype="Int64"),
                   "data1" : np.random.standard_normal(7),
                   "data2" : np.random.standard_normal(7)})

print("DataFrame:")
print(df)
print("\nsize():")
print(df.groupby("key1").size())
print("\ncount():")
print(df.groupby("key1").count())

DataFrame:
   key1  key2     data1     data2
0     a     1  0.728533  0.041276
1     a     2 -0.714949 -1.473617
2  None     1 -1.108375 -0.465517
3     b     2  1.292553 -0.305692
4     b     1  0.020728  1.153793
5     a  <NA> -0.532384  0.302676
6  None     1 -0.751365 -0.295668

size():
key1
a    3
b    2
dtype: int64

count():
      key2  data1  data2
key1                    
a        2      3      3
b        2      2      2


### **Iterating over Groups**
***groupby*** supports ***iteration***, generating ***a sequence of 2-tuples***

In [ ]:
# Display df content
df

key1  key2     data1     data2
0     a     1  0.728533  0.041276
1     a     2 -0.714949 -1.473617
2  None     1 -1.108375 -0.465517
3     b     2  1.292553 -0.305692
4     b     1  0.020728  1.153793
5     a  <NA> -0.532384  0.302676
6  None     1 -0.751365 -0.295668

## **`name` and `grp` is a placeholder**
- holding some information

In [ ]:
# Iteration
# key 1 -> 'a' & 'b'
for name, grp in df.groupby("key1"):
    print(name)
    print(grp)

a
  key1  key2     data1     data2
0    a     1  0.728533  0.041276
1    a     2 -0.714949 -1.473617
5    a  <NA> -0.532384  0.302676
b
  key1  key2     data1     data2
3    b     2  1.292553 -0.305692
4    b     1  0.020728  1.153793


In [ ]:
# Computing a dictionary using one-liner
# for placeholder, I want to make them become dictionary
pieces = {name: group for name, group in df.groupby("key1")}

In [ ]:
# Accessing the key "a" in dictionary
# get back the value
pieces["a"]

key1  key2     data1     data2
0    a     1  0.728533  0.041276
1    a     2 -0.714949 -1.473617
5    a  <NA> -0.532384  0.302676

In [ ]:
# Accessing item in dictionary
pieces["b"]

key1  key2     data1     data2
3    b     2  1.292553 -0.305692
4    b     1  0.020728  1.153793

### **Selecting a Column or Subset of Columns**

In [ ]:
# Returned a DataFrame if a list or array is passed
# If use double square bracket, the output would be a pandas dataframe
df.groupby(["key1", "key2"])[["data2"]].mean()

data2
key1 key2          
a    1     0.041276
     2    -1.473617
b    1     1.153793
     2    -0.305692

In [ ]:
# Returned a grouped Series if only column name is passed as a scalar
# scalar -> a single column name string
df.groupby(["key1", "key2"])["data2"].mean()

key1  key2
a     1       0.041276
      2      -1.473617
b     1       1.153793
      2      -0.305692
Name: data2, dtype: float64

### **Grouping with Dictionaries and Series**

In [ ]:
# Example DataFrame
# Setting the seed number to ensure reproducibility
rng = np.random.default_rng(seed=12345)
people = pd.DataFrame(rng.standard_normal((5, 5)),
                      columns=["a", "b", "c", "d", "e"],
                      index=["Joe", "Steve", "Wanda", "Jill", "Trey"])
people

a         b         c         d         e
Joe   -1.423825  1.263728 -0.870662 -0.259173 -0.075343
Steve -0.740885 -1.367793  0.648893  0.361058 -1.952863
Wanda  2.347410  0.968497 -0.759387  0.902198 -0.466953
Jill  -0.060690  0.788844 -1.256668  0.575858  1.398979
Trey   1.322298 -0.299699  0.902919 -1.621583 -0.158189

In [ ]:
# Add a few NA values using loc function
# loc -> based on label
# iloc -> based on integer
people.loc["Wanda", ["b", "c"]] = np.nan
people

a         b         c         d         e
Joe   -1.423825  1.263728 -0.870662 -0.259173 -0.075343
Steve -0.740885 -1.367793  0.648893  0.361058 -1.952863
Wanda  2.347410       NaN       NaN  0.902198 -0.466953
Jill  -0.060690  0.788844 -1.256668  0.575858  1.398979
Trey   1.322298 -0.299699  0.902919 -1.621583 -0.158189

In [ ]:
# Add a few NA values using iloc function
# single bracket -> indicate a list
# (4, 1) -> refer to rows
# (0, 3) -> refer to columns
people.iloc[(4, 1), (0, 3)] = np.nan
people

a         b         c         d         e
Joe   -1.423825  1.263728 -0.870662 -0.259173 -0.075343
Steve -0.740885 -1.367793  0.648893       NaN -1.952863
Wanda  2.347410       NaN       NaN  0.902198 -0.466953
Jill  -0.060690  0.788844 -1.256668  0.575858  1.398979
Trey        NaN -0.299699  0.902919 -1.621583 -0.158189

In [ ]:
# Group correspondence for the columns
# Act as a dictionary
mapping = {"a": "red", "b": "red", "c": "blue",
           "d": "blue", "e": "red", "f" : "orange", "g" : "violet"}

In [ ]:
# Using the mapping function
# `T` -> transpose
# debugging the code
people.T.groupby(mapping).sum().T

blue       red
Joe   -1.129835 -0.235440
Steve  0.648893 -4.061540
Wanda  0.902198  1.880456
Jill  -0.680811  2.127134
Trey  -0.718663 -0.457888

In [ ]:
# Add the column f and g
rng = np.random.default_rng(seed=12345)
people = people.assign(f = rng.standard_normal((5, 1)),
                       g = rng.standard_normal((5, 1)))
people

a         b         c         d         e         f         g
Joe   -1.423825  1.263728 -0.870662 -0.259173 -0.075343 -1.423825 -0.740885
Steve -0.740885 -1.367793  0.648893       NaN -1.952863  1.263728 -1.367793
Wanda  2.347410       NaN       NaN  0.902198 -0.466953 -0.870662  0.648893
Jill  -0.060690  0.788844 -1.256668  0.575858  1.398979 -0.259173  0.361058
Trey        NaN -0.299699  0.902919 -1.621583 -0.158189 -0.075343 -1.952863

In [ ]:
# Run sum function again
people.T.groupby(mapping).sum().T

blue    orange       red    violet
Joe   -1.129835 -1.423825 -0.235440 -0.740885
Steve  0.648893  1.263728 -4.061540 -1.367793
Wanda  0.902198 -0.870662  1.880456  0.648893
Jill  -0.680811 -0.259173  2.127134  0.361058
Trey  -0.718663 -0.075343 -0.457888 -1.952863

### **Grouping with Functions**

In [ ]:
# Display people content
people

a         b         c         d         e         f         g
Joe   -1.423825  1.263728 -0.870662 -0.259173 -0.075343 -1.423825 -0.740885
Steve -0.740885 -1.367793  0.648893       NaN -1.952863  1.263728 -1.367793
Wanda  2.347410       NaN       NaN  0.902198 -0.466953 -0.870662  0.648893
Jill  -0.060690  0.788844 -1.256668  0.575858  1.398979 -0.259173  0.361058
Trey        NaN -0.299699  0.902919 -1.621583 -0.158189 -0.075343 -1.952863

In [ ]:
# Groupby length of the index letter
people.groupby(len).sum()

a         b         c         d         e         f         g
3 -1.423825  1.263728 -0.870662 -0.259173 -0.075343 -1.423825 -0.740885
4 -0.060690  0.489146 -0.353749 -1.045725  1.240790 -0.334517 -1.591805
5  1.606525 -1.367793  0.648893  0.902198 -2.419816  0.393067 -0.718900

# **13.2 Data Aggregation**
- get some summary data

In [ ]:
# From previous example
df

key1  key2     data1     data2
0     a     1  0.728533  0.041276
1     a     2 -0.714949 -1.473617
2  None     1 -1.108375 -0.465517
3     b     2  1.292553 -0.305692
4     b     1  0.020728  1.153793
5     a  <NA> -0.532384  0.302676
6  None     1 -0.751365 -0.295668

In [ ]:
# Selects the smallest requested number
grouped = df.groupby("key1")
grouped["data1"].nsmallest(2)

key1   
a     1   -0.714949
      5   -0.532384
b     4    0.020728
      3    1.292553
Name: data1, dtype: float64

In [ ]:
# Define own aggregation functions
def peak_2_peak(arr):
    return arr.max() - arr.min()

In [ ]:
# agg: aggregate
# apply the function to the dataframe
grouped.agg(peak_2_peak)

key2     data1     data2
key1                          
a        1  1.443481  1.776293
b        1  1.271825  1.459485

### **`describe()`**: generates **descriptive statistics** for each group in the grouped object.
- provides **a summary of the data**

In [ ]:
# Another way
# describe() -> quick and easy view of our dataset
grouped.describe()

key2                                           data1            ...  \
     count mean       std  min   25%  50%   75%  max count      mean  ...   
key1                                                                  ...   
a      2.0  1.5  0.707107  1.0  1.25  1.5  1.75  2.0   3.0 -0.172933  ...   
b      2.0  1.5  0.707107  1.0  1.25  1.5  1.75  2.0   2.0  0.656640  ...   

                         data2                                          \
           75%       max count      mean       std       min       25%   
key1                                                                     
a     0.098074  0.728533   3.0 -0.376555  0.959031 -1.473617 -0.716170   
b     0.974596  1.292553   2.0  0.424050  1.032012 -0.305692  0.059179   

                                    
           50%       75%       max  
key1                                
a     0.041276  0.171976  0.302676  
b     0.424050  0.788921  1.153793  

[2 rows x 24 columns]

## **Column-Wise and Multiple Function Application**

In [ ]:
# Get the tipping dataset from github
# Download here: https://bit.ly/3VyE0vP
tips = pd.read_csv("https://bit.ly/3VyE0vP")
tips

total_bill   tip smoker   day    time  size
0         16.99  1.01     No   Sun  Dinner     2
1         10.34  1.66     No   Sun  Dinner     3
2         21.01  3.50     No   Sun  Dinner     3
3         23.68  3.31     No   Sun  Dinner     2
4         24.59  3.61     No   Sun  Dinner     4
..          ...   ...    ...   ...     ...   ...
239       29.03  5.92     No   Sat  Dinner     3
240       27.18  2.00    Yes   Sat  Dinner     2
241       22.67  2.00    Yes   Sat  Dinner     2
242       17.82  1.75     No   Sat  Dinner     2
243       18.78  3.00     No  Thur  Dinner     2

[244 rows x 6 columns]

In [ ]:
# Add additional column
tips["tip_pct"] = tips["tip"] / tips['total_bill']
tips

total_bill   tip smoker   day    time  size   tip_pct
0         16.99  1.01     No   Sun  Dinner     2  0.059447
1         10.34  1.66     No   Sun  Dinner     3  0.160542
2         21.01  3.50     No   Sun  Dinner     3  0.166587
3         23.68  3.31     No   Sun  Dinner     2  0.139780
4         24.59  3.61     No   Sun  Dinner     4  0.146808
..          ...   ...    ...   ...     ...   ...       ...
239       29.03  5.92     No   Sat  Dinner     3  0.203927
240       27.18  2.00    Yes   Sat  Dinner     2  0.073584
241       22.67  2.00    Yes   Sat  Dinner     2  0.088222
242       17.82  1.75     No   Sat  Dinner     2  0.098204
243       18.78  3.00     No  Thur  Dinner     2  0.159744

[244 rows x 7 columns]

In [ ]:
# Create an intermediate grouped object
# Group the tips by day and smoker
grouped = tips.groupby(["day", "smoker"])

In [ ]:
# Get descriptive statistics
grouped_pct = grouped["tip_pct"]
grouped_pct.agg("mean")

day   smoker
Fri   No        0.151650
      Yes       0.174783
Sat   No        0.158048
      Yes       0.147906
Sun   No        0.160113
      Yes       0.187250
Thur  No        0.160298
      Yes       0.163863
Name: tip_pct, dtype: float64

## **Difficult to understand the table output without further investigation**

In [ ]:
# Getting back a DataFrame
# Multiple function application
grouped_pct.agg(["mean", "std", peak_2_peak])

mean       std  peak_2_peak
day  smoker                                 
Fri  No      0.151650  0.028123     0.067349
     Yes     0.174783  0.051293     0.159925
Sat  No      0.158048  0.039767     0.235193
     Yes     0.147906  0.061375     0.290095
Sun  No      0.160113  0.042347     0.193226
     Yes     0.187250  0.154134     0.644685
Thur No      0.160298  0.038774     0.193350
     Yes     0.163863  0.039389     0.151240

## **Named aggregation helped to solve this problem**
- by giving **clear column names**

In [ ]:
# Named aggregation (clearer output column names)
grouped_pct.agg(
    mean_tip_pct="mean",
    std_tip_pct="std",
    range=peak_2_peak
)

mean_tip_pct  std_tip_pct     range
day  smoker                                     
Fri  No          0.151650     0.028123  0.067349
     Yes         0.174783     0.051293  0.159925
Sat  No          0.158048     0.039767  0.235193
     Yes         0.147906     0.061375  0.290095
Sun  No          0.160113     0.042347  0.193226
     Yes         0.187250     0.154134  0.644685
Thur No          0.160298     0.038774  0.193350
     Yes         0.163863     0.039389  0.151240

In [ ]:
# Specify a list of functions to apply
# by providing a list
funcs = ["count", "mean", "max", "min", "median"]
result = grouped[["tip_pct", "total_bill"]].agg(funcs)
result

tip_pct                                         total_bill  \
              count      mean       max       min    median      count   
day  smoker                                                              
Fri  No           4  0.151650  0.187735  0.120385  0.149241          4   
     Yes         15  0.174783  0.263480  0.103555  0.173913         15   
Sat  No          45  0.158048  0.291990  0.056797  0.150152         45   
     Yes         42  0.147906  0.325733  0.035638  0.153624         42   
Sun  No          57  0.160113  0.252672  0.059447  0.161665         57   
     Yes         19  0.187250  0.710345  0.065660  0.138122         19   
Thur No          45  0.160298  0.266312  0.072961  0.153492         45   
     Yes         17  0.163863  0.241255  0.090014  0.153846         17   

                                              
                  mean    max    min  median  
day  smoker                                   
Fri  No      18.420000  22.75  12.46  19.235  
     Yes     16.813333  40.17   5.75  13.420  
Sat  No      19.661778  48.33   7.25  17.820  
     Yes     21.276667  50.81   3.07  20.390  
Sun  No      20.506667  48.17   8.77  18.430  
     Yes     24.120000  45.35   7.25  23.100  
Thur No      17.113111  41.19   7.51  15.950  
     Yes     19.190588  43.11  10.34  16.470

In [ ]:
# Extract certain column
result["total_bill"]

count       mean    max    min  median
day  smoker                                        
Fri  No          4  18.420000  22.75  12.46  19.235
     Yes        15  16.813333  40.17   5.75  13.420
Sat  No         45  19.661778  48.33   7.25  17.820
     Yes        42  21.276667  50.81   3.07  20.390
Sun  No         57  20.506667  48.17   8.77  18.430
     Yes        19  24.120000  45.35   7.25  23.100
Thur No         45  17.113111  41.19   7.51  15.950
     Yes        17  19.190588  43.11  10.34  16.470

## **Using tuple**
- non-immutable
- **it has an sequential order**

In [ ]:
# Passing a list of tuples
ftuples = [("Average", "mean"), ("Variation", "var")]
grouped[["tip_pct", "total_bill"]].agg(ftuples)

tip_pct           total_bill            
              Average Variation    Average   Variation
day  smoker                                           
Fri  No      0.151650  0.000791  18.420000   25.596333
     Yes     0.174783  0.002631  16.813333   82.562438
Sat  No      0.158048  0.001581  19.661778   79.908965
     Yes     0.147906  0.003767  21.276667  101.387535
Sun  No      0.160113  0.001793  20.506667   66.099980
     Yes     0.187250  0.023757  24.120000  109.046044
Thur No      0.160298  0.001503  17.113111   59.625081
     Yes     0.163863  0.001551  19.190588   69.808518

In [ ]:
# Original code
# very lengthy and hard to "see"
grouped.agg(tip_max=('tip', 'max'), size_sum=('size', 'sum'))

tip_max  size_sum
day  smoker                   
Fri  No         3.50         9
     Yes        4.73        31
Sat  No         9.00       115
     Yes       10.00       104
Sun  No         6.00       167
     Yes        6.50        49
Thur No         6.70       112
     Yes        5.00        40

In [ ]:
# Implicit line continuation
# Wrap the code in parenthesis -> "()"
# Any code wrapped in parenthesis, can be broken into different rows
(
    grouped
    .agg(tip_max=('tip', 'max'),
         size_sum=('size', 'sum'))
)

tip_max  size_sum
day  smoker                   
Fri  No         3.50         9
     Yes        4.73        31
Sat  No         9.00       115
     Yes       10.00       104
Sun  No         6.00       167
     Yes        6.50        49
Thur No         6.70       112
     Yes        5.00        40

In [ ]:
# Another fancy way
grouped.agg({"total_bill" : ["min", "max", "mean", "std"],
             "size" : "sum"})

total_bill                              size
                   min    max       mean        std  sum
day  smoker                                             
Fri  No          12.46  22.75  18.420000   5.059282    9
     Yes          5.75  40.17  16.813333   9.086388   31
Sat  No           7.25  48.33  19.661778   8.939181  115
     Yes          3.07  50.81  21.276667  10.069138  104
Sun  No           8.77  48.17  20.506667   8.130189  167
     Yes          7.25  45.35  24.120000  10.442511   49
Thur No           7.51  41.19  17.113111   7.721728  112
     Yes         10.34  43.11  19.190588   8.355149   40

In [ ]:
# Implicit line continuation
(
    grouped[["tip_pct", "total_bill"]]
    .agg(["count", "mean", "median"])
    .rename_axis(index=["day", "smoker"])
)


tip_pct                     total_bill                   
              count      mean    median      count       mean  median
day  smoker                                                          
Fri  No           4  0.151650  0.149241          4  18.420000  19.235
     Yes         15  0.174783  0.173913         15  16.813333  13.420
Sat  No          45  0.158048  0.150152         45  19.661778  17.820
     Yes         42  0.147906  0.153624         42  21.276667  20.390
Sun  No          57  0.160113  0.161665         57  20.506667  18.430
     Yes         19  0.187250  0.138122         19  24.120000  23.100
Thur No          45  0.160298  0.153492         45  17.113111  15.950
     Yes         17  0.163863  0.153846         17  19.190588  16.470

## **Returning Aggregated Data Without Row Indexes**
- Using **`as_index`=False/True**

In [ ]:
# Disable index
(tips
 .groupby(["day", "smoker"], as_index=False)
 [['total_bill', 'tip', 'size', 'tip_pct']]
 .mean()
)

day smoker  total_bill       tip      size   tip_pct
0   Fri     No   18.420000  2.812500  2.250000  0.151650
1   Fri    Yes   16.813333  2.714000  2.066667  0.174783
2   Sat     No   19.661778  3.102889  2.555556  0.158048
3   Sat    Yes   21.276667  2.875476  2.476190  0.147906
4   Sun     No   20.506667  3.167895  2.929825  0.160113
5   Sun    Yes   24.120000  3.516842  2.578947  0.187250
6  Thur     No   17.113111  2.673778  2.488889  0.160298
7  Thur    Yes   19.190588  3.030000  2.352941  0.163863

# **13.3 Apply: General split-apply-combine**

In [ ]:
# Define our own "top" function call
# Define a function to select the top five tip_pct values by group
def top(df, n=5, column="tip_pct"):
    return df.sort_values(column, ascending=False)[:n]

# Overwrite the default value
top(tips, n=3)

total_bill   tip smoker  day    time  size   tip_pct
172        7.25  5.15    Yes  Sun  Dinner     2  0.710345
178        9.60  4.00    Yes  Sun  Dinner     2  0.416667
67         3.07  1.00    Yes  Sat  Dinner     1  0.325733

## Using our own function has **one caveat**
- in terms of **speed**, may be it is **not that "efficient"**
- compared to **native python code**

In [ ]:
# Direct chaining
# instead of using the function "top"
# we use head() -> native python function
tips.sort_values("tip_pct", ascending=False).head(3)

total_bill   tip smoker  day    time  size   tip_pct
172        7.25  5.15    Yes  Sun  Dinner     2  0.710345
178        9.60  4.00    Yes  Sun  Dinner     2  0.416667
67         3.07  1.00    Yes  Sat  Dinner     1  0.325733

## **Benchmarking code**
We will **almost certainly** observe (**in terms of execution speed**):

`Direct chaining`  <  `Function version`

In [ ]:
import timeit

# Number of repetitions
n_runs = 1000

# Timing Code A (function call)
time_func = timeit.timeit(
    stmt="top(tips, n=6)",
    globals=globals(),
    number=n_runs
)

# Timing Code B (direct chaining)
time_direct = timeit.timeit(
    stmt='tips.sort_values("tip_pct", ascending=False).head(6)',
    globals=globals(),
    number=n_runs
)

print(f"Function version: {time_func:.6f} seconds")
print(f"Direct chaining: {time_direct:.6f} seconds")


Function version: 0.246148 seconds
Direct chaining: 0.245763 seconds


In [ ]:
# display tips dataframe
tips

total_bill   tip smoker   day    time  size   tip_pct
0         16.99  1.01     No   Sun  Dinner     2  0.059447
1         10.34  1.66     No   Sun  Dinner     3  0.160542
2         21.01  3.50     No   Sun  Dinner     3  0.166587
3         23.68  3.31     No   Sun  Dinner     2  0.139780
4         24.59  3.61     No   Sun  Dinner     4  0.146808
..          ...   ...    ...   ...     ...   ...       ...
239       29.03  5.92     No   Sat  Dinner     3  0.203927
240       27.18  2.00    Yes   Sat  Dinner     2  0.073584
241       22.67  2.00    Yes   Sat  Dinner     2  0.088222
242       17.82  1.75     No   Sat  Dinner     2  0.098204
243       18.78  3.00     No  Thur  Dinner     2  0.159744

[244 rows x 7 columns]

In [ ]:
# Apply function on certain selected columns
tips.groupby("smoker")[['total_bill', 'tip', 'size', 'tip_pct']].apply(top)

total_bill   tip  size   tip_pct
smoker                                      
No     232       11.61  3.39     2  0.291990
       149        7.51  2.00     2  0.266312
       51        10.29  2.60     2  0.252672
       185       20.69  5.00     5  0.241663
       88        24.71  5.85     2  0.236746
Yes    172        7.25  5.15     2  0.710345
       178        9.60  4.00     2  0.416667
       67         3.07  1.00     1  0.325733
       183       23.17  6.50     4  0.280535
       109       14.31  4.00     2  0.279525

In [ ]:
# Overwrite the default values
(
    tips
    .groupby("smoker")[['total_bill', 'tip', 'size', 'tip_pct']]
    .apply(top, n=3, column="total_bill")
)

total_bill    tip  size   tip_pct
smoker                                       
No     212       48.33   9.00     4  0.186220
       59        48.27   6.73     4  0.139424
       156       48.17   5.00     6  0.103799
Yes    170       50.81  10.00     3  0.196812
       182       45.35   3.50     3  0.077178
       102       44.30   2.50     3  0.056433

In [ ]:
# Earlier example
result = tips.groupby("smoker")["tip_pct"].describe()
result

count      mean       std       min       25%       50%       75%  \
smoker                                                                      
No      151.0  0.159328  0.039910  0.056797  0.136906  0.155625  0.185014   
Yes      93.0  0.163196  0.085119  0.035638  0.106771  0.153846  0.195059   

             max  
smoker            
No      0.291990  
Yes     0.710345

In [ ]:
# Unstack() -> from wide to vertical
result.unstack("smoker")

smoker
count  No        151.000000
       Yes        93.000000
mean   No          0.159328
       Yes         0.163196
std    No          0.039910
       Yes         0.085119
min    No          0.056797
       Yes         0.035638
25%    No          0.136906
       Yes         0.106771
50%    No          0.155625
       Yes         0.153846
75%    No          0.185014
       Yes         0.195059
max    No          0.291990
       Yes         0.710345
dtype: float64

## **Filling Missing Values with Group-Specific Values**

In [ ]:
# Create a mock data
rng = np.random.default_rng(seed=12345)
s =pd.Series(rng.standard_normal(6))
s

0   -1.423825
1    1.263728
2   -0.870662
3   -0.259173
4   -0.075343
5   -0.740885
dtype: float64

In [ ]:
# Skip of 2
s.iloc[::2] = np.nan
s

0         NaN
1    1.263728
2         NaN
3   -0.259173
4         NaN
5   -0.740885
dtype: float64

In [ ]:
# Fill NA with mean
s.fillna(s.mean())

0    0.087890
1    1.263728
2    0.087890
3   -0.259173
4    0.087890
5   -0.740885
dtype: float64

In [ ]:
# Another example data
states = ["Ohio", "New York", "Vermont", "Florida",
          "Oregon", "Nevada", "California", "Idaho"]
group_key = ["East", "East", "East", "East",
             "West", "West", "West", "West"]
data = pd.Series(rng.standard_normal(8), index=states)
data

Ohio         -1.367793
New York      0.648893
Vermont       0.361058
Florida      -1.952863
Oregon        2.347410
Nevada        0.968497
California   -0.759387
Idaho         0.902198
dtype: float64

In [ ]:
# Insert NaN value into the series
data[["Vermont", "Nevada", "Idaho"]] = np.nan
data

Ohio         -1.367793
New York      0.648893
Vermont            NaN
Florida      -1.952863
Oregon        2.347410
Nevada             NaN
California   -0.759387
Idaho              NaN
dtype: float64

In [ ]:
# Extract size
data.groupby(group_key).size()

East    4
West    4
dtype: int64

In [ ]:
# Using count function
# count() function deals only with non-null value
# will ignore Nan value
data.groupby(group_key).count()

East    3
West    2
dtype: int64

In [ ]:
# Original shape
data

Ohio         -1.367793
New York      0.648893
Vermont            NaN
Florida      -1.952863
Oregon        2.347410
Nevada             NaN
California   -0.759387
Idaho              NaN
dtype: float64

## This is using **`transform`** function
- `transform` function **retained the original shape** of dataset

In [ ]:
# Fill NA using group mean
# preferred: transform is vectorized and aligns to original rows
data = data.fillna(
    data.groupby(group_key).transform("mean")
)
data

Ohio         -1.367793
New York      0.648893
Vermont      -0.890588
Florida      -1.952863
Oregon        2.347410
Nevada        0.794011
California   -0.759387
Idaho         0.794011
dtype: float64

## This is using **`apply`** function
- will **return as much information as possible**

In [ ]:
# Fill the NA values using the group means
def fill_mean(group):
    return group.fillna(group.mean())

data.groupby(group_key).apply(fill_mean)

East  Ohio         -1.367793
      New York      0.648893
      Vermont      -0.890588
      Florida      -1.952863
West  Oregon        2.347410
      Nevada        0.794011
      California   -0.759387
      Idaho         0.794011
dtype: float64

## **Caveat here: `transform` vs `apply`**
- in terms of **execution speed**: **`transform`** < **`apply`**

In [ ]:
# Another example data
states = ["Ohio", "New York", "Vermont", "Florida",
          "Oregon", "Nevada", "California", "Idaho"]
group_key = ["East", "East", "East", "East",
             "West", "West", "West", "West"]
data = pd.Series(rng.standard_normal(8), index=states)
data

Ohio         -0.466953
New York     -0.060690
Vermont       0.788844
Florida      -1.256668
Oregon        0.575858
Nevada        1.398979
California    1.322298
Idaho        -0.299699
dtype: float64

In [ ]:
# Insert NaN value into the series
data[["Vermont", "Nevada", "Idaho"]] = np.nan
data

Ohio         -0.466953
New York     -0.060690
Vermont            NaN
Florida      -1.256668
Oregon        0.575858
Nevada             NaN
California    1.322298
Idaho              NaN
dtype: float64

In [ ]:
# With predefined fill values
fill_values = {"East": 0.5, "West": -1}

def fill_func(group):
    return group.fillna(fill_values[group.name])

In [ ]:
# Filling in the NA with predefined value
data.groupby(group_key).apply(fill_func)

East  Ohio         -0.466953
      New York     -0.060690
      Vermont       0.500000
      Florida      -1.256668
West  Oregon        0.575858
      Nevada       -1.000000
      California    1.322298
      Idaho        -1.000000
dtype: float64

# **13.4 Group Transforms and “Unwrapped” GroupBys**
- looking at **in terms of execution speed**

In [ ]:
# An example data
df = pd.DataFrame({'key': ['a', 'b', 'c'] * 4,
                   'value': np.arange(12.)})
df

key  value
0    a    0.0
1    b    1.0
2    c    2.0
3    a    3.0
4    b    4.0
5    c    5.0
6    a    6.0
7    b    7.0
8    c    8.0
9    a    9.0
10   b   10.0
11   c   11.0

In [ ]:
# Group means by key
# Create an intermediate grouped object
g = df.groupby('key')['value']
g.mean()

key
a    4.5
b    5.5
c    6.5
Name: value, dtype: float64

In [ ]:
# Define a mean function
def get_mean(g):
    return g.mean()

In [ ]:
# Transform into a Series of mean data
# try to get back the original shape
g.transform(get_mean)

0     4.5
1     5.5
2     6.5
3     4.5
4     5.5
5     6.5
6     4.5
7     5.5
8     6.5
9     4.5
10    5.5
11    6.5
Name: value, dtype: float64

In [ ]:
# Another way - this is faster using default function
g.transform('mean')

0     4.5
1     5.5
2     6.5
3     4.5
4     5.5
5     6.5
6     4.5
7     5.5
8     6.5
9     4.5
10    5.5
11    6.5
Name: value, dtype: float64

In [ ]:
# A normalization function
def normalize(x):
    return (x - x.mean()) / x.std()

In [ ]:
# Method 1 - faster
g.transform(normalize)

0    -1.161895
1    -1.161895
2    -1.161895
3    -0.387298
4    -0.387298
5    -0.387298
6     0.387298
7     0.387298
8     0.387298
9     1.161895
10    1.161895
11    1.161895
Name: value, dtype: float64

In [ ]:
# Method 2: apply - slower
# index of the resulting Series will be a MultiIndex
g.apply(normalize)

key    
a    0    -1.161895
     3    -0.387298
     6     0.387298
     9     1.161895
b    1    -1.161895
     4    -0.387298
     7     0.387298
     10    1.161895
c    2    -1.161895
     5    -0.387298
     8     0.387298
     11    1.161895
Name: value, dtype: float64

## **Unwrapped function**
- straight away **without intermediate grouped object**
- almost certainly **the fastest** compared to `transform` and `apply`

In [ ]:
# Unwrapped group operations - doing arithmetic between the outputs of
# multiple GroupBy operations
# often much faster than a general apply function
normalized = (df['value'] - g.transform('mean')) / g.transform('std')
normalized

0    -1.161895
1    -1.161895
2    -1.161895
3    -0.387298
4    -0.387298
5    -0.387298
6     0.387298
7     0.387298
8     0.387298
9     1.161895
10    1.161895
11    1.161895
Name: value, dtype: float64

# **13.5 Pivot Tables and Cross-Tabulation**
A ***pivot table*** is a ***data summarization*** tool
- like Microsoft Excel

In [ ]:
# From previous tipping dataset
tips

total_bill   tip smoker   day    time  size   tip_pct
0         16.99  1.01     No   Sun  Dinner     2  0.059447
1         10.34  1.66     No   Sun  Dinner     3  0.160542
2         21.01  3.50     No   Sun  Dinner     3  0.166587
3         23.68  3.31     No   Sun  Dinner     2  0.139780
4         24.59  3.61     No   Sun  Dinner     4  0.146808
..          ...   ...    ...   ...     ...   ...       ...
239       29.03  5.92     No   Sat  Dinner     3  0.203927
240       27.18  2.00    Yes   Sat  Dinner     2  0.073584
241       22.67  2.00    Yes   Sat  Dinner     2  0.088222
242       17.82  1.75     No   Sat  Dinner     2  0.098204
243       18.78  3.00     No  Thur  Dinner     2  0.159744

[244 rows x 7 columns]

## **Pivot table: three parameters**
1. `index`,
2. `columns`,
3. `values`

In [ ]:
# Getting mean for size and tip_pct columns
tips.pivot_table(index=["time", "day"], columns="smoker",
                 values=["tip_pct", "size"])

size             tip_pct          
smoker             No       Yes        No       Yes
time   day                                         
Dinner Fri   2.000000  2.222222  0.139622  0.165347
       Sat   2.555556  2.476190  0.158048  0.147906
       Sun   2.929825  2.578947  0.160113  0.187250
       Thur  2.000000       NaN  0.159744       NaN
Lunch  Fri   3.000000  1.833333  0.187735  0.188937
       Thur  2.500000  2.352941  0.160311  0.163863

In [ ]:
# Use aggfunc = len for count
tips.pivot_table(index=["time", "smoker"], columns="day",
                 values="tip_pct", aggfunc=len, margins=True)

day             Fri   Sat   Sun  Thur  All
time   smoker                             
Dinner No       3.0  45.0  57.0   1.0  106
       Yes      9.0  42.0  19.0   NaN   70
Lunch  No       1.0   NaN   NaN  44.0   45
       Yes      6.0   NaN   NaN  17.0   23
All            19.0  87.0  76.0  62.0  244

In [ ]:
# pass NA with a specific value using fill_value
# placeholder / sentinel value
tips.pivot_table(index=["time", "size", "smoker"], columns="day",
                 values="tip_pct", fill_value=-999)

day                        Fri         Sat         Sun        Thur
time   size smoker                                                
Dinner 1    No     -999.000000    0.137931 -999.000000 -999.000000
            Yes    -999.000000    0.325733 -999.000000 -999.000000
       2    No        0.139622    0.162705    0.168859    0.159744
            Yes       0.171297    0.148668    0.207893 -999.000000
       3    No     -999.000000    0.154661    0.152663 -999.000000
            Yes    -999.000000    0.144995    0.152660 -999.000000
       4    No     -999.000000    0.150096    0.148143 -999.000000
            Yes       0.117750    0.124515    0.193370 -999.000000
       5    No     -999.000000 -999.000000    0.206928 -999.000000
            Yes    -999.000000    0.106572    0.065660 -999.000000
       6    No     -999.000000 -999.000000    0.103799 -999.000000
Lunch  1    No     -999.000000 -999.000000 -999.000000    0.181728
            Yes       0.223776 -999.000000 -999.000000 -999.000000
       2    No     -999.000000 -999.000000 -999.000000    0.166005
            Yes       0.181969 -999.000000 -999.000000    0.158843
       3    No        0.187735 -999.000000 -999.000000    0.084246
            Yes    -999.000000 -999.000000 -999.000000    0.204952
       4    No     -999.000000 -999.000000 -999.000000    0.138919
            Yes    -999.000000 -999.000000 -999.000000    0.155410
       5    No     -999.000000 -999.000000 -999.000000    0.121389
       6    No     -999.000000 -999.000000 -999.000000    0.173706

## **Cross-Tabulations: Crosstab**
A ***special pivot table*** to compute ***group frequencies***

In [ ]:
from io import StringIO

In [ ]:
# Data consists of strings
data = """Sample  Nationality  Handedness
1   USA  Right-handed
2   Japan    Left-handed
3   USA  Right-handed
4   Japan    Right-handed
5   Japan    Left-handed
6   Japan    Right-handed
7   USA  Right-handed
8   USA  Left-handed
9   Japan    Right-handed
10  USA  Right-handed"""

### **`sep=r"\s+"`**
- "one or more whitespace characters"

### also `sep=\t, csv`

In [ ]:
# Create a table
data = pd.read_table(StringIO(data), sep=r"\s+")

In [ ]:
# Display the data content
data

Sample Nationality    Handedness
0       1         USA  Right-handed
1       2       Japan   Left-handed
2       3         USA  Right-handed
3       4       Japan  Right-handed
4       5       Japan   Left-handed
5       6       Japan  Right-handed
6       7         USA  Right-handed
7       8         USA   Left-handed
8       9       Japan  Right-handed
9      10         USA  Right-handed

In [ ]:
# Crosstab Example 1 - find group frequencies
# Summarize by nationality and handedness
# Crosstab: row and column
pd.crosstab(data["Nationality"], data["Handedness"], margins=True)

Handedness   Left-handed  Right-handed  All
Nationality                                
Japan                  2             3    5
USA                    1             4    5
All                    3             7   10

In [ ]:
# Tipping dataset
tips

total_bill   tip smoker   day    time  size   tip_pct
0         16.99  1.01     No   Sun  Dinner     2  0.059447
1         10.34  1.66     No   Sun  Dinner     3  0.160542
2         21.01  3.50     No   Sun  Dinner     3  0.166587
3         23.68  3.31     No   Sun  Dinner     2  0.139780
4         24.59  3.61     No   Sun  Dinner     4  0.146808
..          ...   ...    ...   ...     ...   ...       ...
239       29.03  5.92     No   Sat  Dinner     3  0.203927
240       27.18  2.00    Yes   Sat  Dinner     2  0.073584
241       22.67  2.00    Yes   Sat  Dinner     2  0.088222
242       17.82  1.75     No   Sat  Dinner     2  0.098204
243       18.78  3.00     No  Thur  Dinner     2  0.159744

[244 rows x 7 columns]

In [ ]:
# Crosstab Example 2
pd.crosstab([tips["time"], tips["day"]], tips["smoker"], margins=True)

smoker        No  Yes  All
time   day                
Dinner Fri     3    9   12
       Sat    45   42   87
       Sun    57   19   76
       Thur    1    0    1
Lunch  Fri     1    6    7
       Thur   44   17   61
All          151   93  244

# **Closing Recap: Choosing the Right Tool in Grouped Analysis**

### When to use `agg` vs `transform` vs `apply`

**Use `agg()`** when you want **one row per group** (summary tables, reporting).
- Fast and optimized.
- Best for group-level statistics.

**Use `transform()`** when you want **the same number of rows as the original data** (group-wise calculations).
- Ideal for normalization, filling NA by group, ranking within group.
- Vectorized and typically faster than `apply()`.

**Use `apply()`** only when the logic **cannot be expressed** using `agg()` or `transform()`.
- Most flexible, but usually slower.
- Prefer avoiding it for standard tasks like top-N or filling missing values.

---

# **When to use `pivot_table()` vs `groupby().agg().unstack()`**

**Use `pivot_table()`** for quick, Excel-style summaries and teaching-friendly cross-tab summaries.
- Concise syntax.
- Handles missing combinations cleanly.

**Use `groupby().agg().unstack()`** when you want more explicit, step-by-step control.
- Better for complex pipelines and debugging.
- More transparent for production code.
- wide to vertical table, and vice versa


# **Revision -> recreate this notebook using Jupyter Notebook**
- **important for answering Final exam question**

# **That's all for the day :)**